### ISSUE TITLE
Broken pipe if monitor down

### ISSUE THREAD

**Issue by casselt** [Crapworks/ceph-dash#52]
Hi,

first of all thanks for the great work, I love the REST API.

I run a cluster with three nodes and three monitors. If I restart a monitor for maintenance, ceph-dash hangs and shows empty output on the website. It does not happen for every monitor. It seems that ceph-dash is picking a “main” monitor and has problems switching over if this one fails. REST API is responding though, but with a huge delay, I measured between 20 and 25 seconds.
Example: I restart monitor 3, everything's fine. After monitor 3 is up again I restart monitor 2, ceph-dash hangs. If I restart monitor 2 again, everything's fine, ceph-dash seems to have switched to another monitor. Restarting monitor 1 after that leads to a freeze again, so ceph-dash switched to this monitor apparently.

Output of ceph-dash.py:
```
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: 127.0.0.1 - - [06/Jun/2017 09:02:20] "GET /ceph-dash HTTP/1.0" 200 -
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: Traceback (most recent call last):
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: File "/usr/lib/python2.7/SocketServer.py", line 295, in _handle_request_noblock
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: self.process_request(request, client_address)
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: File "/usr/lib/python2.7/SocketServer.py", line 321, in process_request
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: self.finish_request(request, client_address)
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: File "/usr/lib/python2.7/SocketServer.py", line 334, in finish_request
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: self.RequestHandlerClass(request, client_address, self)
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: File "/usr/lib/python2.7/SocketServer.py", line 657, in __init__
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: self.finish()
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: File "/usr/lib/python2.7/SocketServer.py", line 716, in finish
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: self.wfile.close()
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: File "/usr/lib/python2.7/socket.py", line 279, in close
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: self.flush()
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: File "/usr/lib/python2.7/socket.py", line 303, in flush
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: self._sock.sendall(view[write_offset:write_offset+buffer_size])
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: error: [Errno 32] Broken pipe
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: ----------------------------------------
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: Exception happened during processing of request from ('127.0.0.1', 59079)
Jun 06 09:02:20 ceph-log ceph-dash.py[27784]: ----------------------------------------
```

**Comment by Crapworks** [comment]
Hi @casselt,

I am not sure that this is an issue with ceph-dash. The dashboard basically issues a `ceph status --format=json`. If that doesn't work correctly (for example is hanging or taking long) then ceph-dash of course also has those problems. Can you issue the mentioned command on the same node you are running your dashboard and check if this command also hangs or has the same problems?

**Comment by casselt** [comment]
Hi @Crapworks,

thanks for the fast answer. I can't issue this command because I don't installed ceph on this server. It's only for logging. I installed python-ceph which is listed as the only requirement (plus flask of course). I use the python-ceph packaged in debian jessie, so it might be a problem of outdated software. I'll try to test with a newer version and report back.

**Comment by Crapworks** [comment]
As a workaround you can issue the command on one of the monitors as well. There shouldn't be a huge difference. Just to see if the command is hanging under the mentioned conditions.

**Comment by casselt** [comment]
I issued the command on the two other monitor nodes plus a management node. Works without a problem.
But my guess was correct, the python-ceph package in debian jessie was too old it seems. I upgraded to the jessie-backports version and only a minor hiccup for about 20 seconds with the error above, after that normal operation.
Thanks for the fast answer and the hint to “ceph status”.

Have a great day!
